FRENCH/src/ch12-04-testing-the-librarys-functionality.md

```diff
@@ -0,0 +1,644 @@
+vous avez vu cette façon d'utiliser un itérateur dans
+[l'encart 3-5][ch3-iter]<!-- ignore -->, dans lequel nous avions utilisé une
+boucle `for` sur un itérateur pour exécuter du code sur chaque élément d'une
+collection.
+
+<!--
+#### Searching Each Line for the Query
+-->
+
+#### Trouver chaque ligne correspondante à la recherche
+
+<!--
+Next, we’ll check whether the current line contains our query string.
+Fortunately, strings have a helpful method named `contains` that does this for
+us! Add a call to the `contains` method in the `search` function, as shown in
+Listing 12-18. Note this still won’t compile yet.
+-->
+
+Ensuite, nous allons vérifier que la ligne courante contient la chaîne de
+caractères que nous recherchons. Heureusement, les chaînes de caractères ont une
+méthode `contains` assez pratique qui fait cela pour nous ! Ajoutez l'appel à
+la méthode `contains` dans la fonction `rechercher`, comme dans l'encart 12-18.
+Notez qu'ici non plus nous ne pouvons pas encore compiler.
+
+<!--
+<span class="filename">Filename: src/lib.rs</span>
+-->
+
+<span class="filename">Fichier : src/lib.rs</span>
+
+<!--
+```rust,ignore
+{{#rustdoc_include ../listings-sources/ch12-an-io-project/listing-12-18/src/lib.rs:here}}
+```
+-->
+
+```rust,ignore
+{{#rustdoc_include ../listings/ch12-an-io-project/listing-12-18/src/lib.rs:here}}
+```
+
+<!--
+<span class="caption">Listing 12-18: Adding functionality to see whether the
+line contains the string in `query`</span>
+-->
+
+<span class="caption">Encart 12-18 : Ajout d'une fonctionnalité pour trouver
+quelle ligne contient la chaîne de caractères `recherche`</span>
+
+<!--
+#### Storing Matching Lines
+-->
+
+#### Stocker les lignes trouvées
+
+<!--
+We also need a way to store the lines that contain our query string. For that,
+we can make a mutable vector before the `for` loop and call the `push` method
+to store a `line` in the vector. After the `for` loop, we return the vector, as
+shown in Listing 12-19.
+-->
+
+Nous avons aussi besoin d'un moyen de stocker les lignes qui contienent la
+chaîne de caractères que nous recherchons. Pour cela, nous pouvons créer un
+vecteur mutable avant la boucle `for` et appeler la méthode `push` pour
+enregistrer la `ligne` dans le vecteur. Après la boucle `for`, nous retournons
+le vecteur, comme dans l'encart 12-19 :
+
+<!--
+<span class="filename">Filename: src/lib.rs</span>
+-->
+
+<span class="filename">Fichier : src/lib.rs</span>
+
+<!--
+```rust,ignore
+{{#rustdoc_include ../listings-sources/ch12-an-io-project/listing-12-19/src/lib.rs:here}}
+```
+-->
+
+```rust,ignore
+{{#rustdoc_include ../listings/ch12-an-io-project/listing-12-19/src/lib.rs:here}}
+```
+
+```rust,ignore
+pub fn rechercher<'a>(recherche: &str, contenu: &'a str) -> Vec<&'a str> {
+    let mut resultats = Vec::new();
+
+    for ligne in contenu.lines() {
+        if ligne.contains(recherche) {
+            resultats.push(ligne);
+        }
+    }
+
+    resultats
+}
+```
```

Comment:
```suggestion
```

_Linked to @yozhgoor suggestion in https://github.com/Jimskapt/rust-book-fr/pull/139#r674042888_